olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp

```diff
@@ -662,6 +663,68 @@ VersionedLayerClientImpl::GetAggregatedData(TileRequest request) {
   return client::CancellableFuture<AggregatedDataResponse>(
       std::move(cancel_token), std::move(promise));
 }
+
+client::CancellationToken VersionedLayerClientImpl::Protect(
+    const TileKeys& tiles, ProtectCallback callback) {
+  auto version = catalog_version_.load();
+  auto protect_task =
+      [=](client::CancellationContext context) -> ProtectResponse {
+    // could protect keys, which is already in cache
+    // because we don't know quad tree which will contain tiles
+    repository::DataCacheRepository data_cache_repository(catalog_,
+                                                          settings_.cache);
+    repository::PartitionsCacheRepository partitions_cache_repository(
+        catalog_, settings_.cache);
+
+    cache::KeyValueCache::KeyListType keys_to_protect;
+    keys_to_protect.reserve(tiles.size() * 2);
+
+    std::vector<read::QuadTreeIndex> quad_trees;
+    quad_trees.reserve(tiles.size());
+    for (const auto& tile : tiles) {
+      auto it = std::find_if(
+          quad_trees.begin(), quad_trees.end(),
+          [&tile](const read::QuadTreeIndex& quad) {
+            auto root = quad.GetRootTile();
+            return (tile.Level() >= root.Level() &&
+                    tile.Level() - root.Level() <= kQuadTreeDepth &&
+                    root.IsParentOf(tile));
+          });
+      if (it == quad_trees.end()) {
+        quad_trees.emplace_back();
+        read::QuadTreeIndex& cached_tree = quad_trees.back();
```

Comment:
Better move instead of creating an empty element and never removing it when there is no such quad tree.